xen/arch/arm/coproc/schedule.c

```diff
@@ -0,0 +1,317 @@
+/*
+ * xen/arch/arm/coproc/schedule.c
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <xen/init.h>
+#include <xen/err.h>
+
+#include "coproc.h"
+
+/* TODO Should be coproc specific */
+static s_time_t coproc_wait_time = MILLISECS(500);
+
+/* TODO Each coproc may have its own algorithm */
+static char __initdata opt_vcoproc_sched[10] = "";
+string_param("vcoproc_sched", opt_vcoproc_sched);
+
+static struct vcoproc_scheduler *const vcoproc_schedulers[] = {
+	NULL
+};
+
+#define VCOPROC_SCHED_OP(opsptr, fn, ...)                                  \
+         (( (opsptr)->fn != NULL ) ? (opsptr)->fn(opsptr, ##__VA_ARGS__ )  \
+          : (typeof((opsptr)->fn(opsptr, ##__VA_ARGS__)))0 )
+
+static void vcoproc_scheduler_set_current(const struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+
+	sched_data->curr = vcoproc;
+}
+
+static struct vcoproc_instance *vcoproc_scheduler_get_current(const struct vcoproc_scheduler *sched)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+
+	return sched_data->curr;
+}
+
+bool_t vcoproc_scheduler_vcoproc_is_destroyed(struct vcoproc_scheduler *sched,
+		struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+	bool_t destroyed;
+
+	if ( !vcoproc )
+		return true;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	destroyed = vcoproc->state == VCOPROC_UNKNOWN ? true : false;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return destroyed;
+}
+
+int vcoproc_scheduler_vcoproc_init(struct vcoproc_scheduler *sched,
+		struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	if ( !vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc) )
+		return -EINVAL;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	vcoproc->sched_priv = VCOPROC_SCHED_OP(sched, alloc_vdata, vcoproc);
+	if ( !vcoproc->sched_priv )
+	{
+		printk("Failed to allocate scheduler-specific data for vcoproc \"%s\"\n",
+				dev_path(vcoproc->coproc->dev));
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return -ENOMEM;
+	}
+
+	vcoproc->state = VCOPROC_SLEEPING;
```

Comment:
vcoproc->state setting seems to be the only thing here to be protected by spinlock, if anything